### PR TITLE
Add instructions for debugging via VS Code

### PR DESCRIPTION
**Pull request by fgh1999**
Reopen #569.

### CODE REVIEW

File: docs/src/kernel/advanced-instructions.md [hunk 1]
```diff
@@ -60,16 +60,31 @@ Then, in the interactive shell, run the following script to start the syscall te
...
+Start a GDB-enabled VM of Asterinas with `GDB_SERVER=1` and wait for debugging connection:
 
 ```bash
 make run GDB_SERVER=1 ENABLE_KVM=0
 ```
 
-Next, start a GDB client in another terminal.
+Then debug via [a GDB client in terminal](#terminal), or [VS Code](#vs-code).
+
+Currently, the Asterinas runner's debugging interface is exposed via UNIX sockets. Thus there shouldn't be multiple debugging instances in the same container. To add debug symbols for the underlying infrastructures such as UEFI firmware or bootloader, please check the runner's source code for details.
```
Comment (tatetian) [inline docs/src/kernel/advanced-instructions.md:71]: Follow semantic line breaks like the rest of the doc.

File: docs/src/kernel/advanced-instructions.md [hunk 2]
```diff
@@ -60,16 +60,31 @@ Then, in the interactive shell, run the following script to start the syscall te
...
 
-Currently, the Asterinas runner's debugging interface is exposed via UNIX sockets. Thus there shouldn't be multiple debugging instances in the same container. To add debug symbols for the underlying infrastructures such as UEFI firmware or bootloader, please check the runner's source code for details.
+#### VS Code
+
+[CodeLLDB](https://marketplace.visualstudio.com/items?itemName=vadimcn.vscode-lldb) extension is required.
+
+After starting a debug server with `GDB_SERVER=1` from the shell as [described above](#using-gdb-to-debug), press `F5`(Run and Debug) to start a debug session via VS Code. 
+Now click `Continue`(or, press `F5`) at the fisrt break to resume the paused server instance, and it will continue until reaching your first breakpoint. 
+
+![GDB Debugging via VS Code](../images/gdb_debug_via_vscode.png)
```
Comment (tatetian) [inline docs/src/kernel/advanced-instructions.md:88]: Binary files like PNG are not friendly to version control. My current strategy is (1) to be very restrained with adding image files into the Git history, (2) if adding the images is unavoidable, then prefer vector image format like SVG. Vector format usually means much smaller file sizes.

I don't know how essential is this screenshot. If it is not essential, just don't add the image. Otherwise, use [Git LFS](https://git-lfs.com/) for more efficient version control of binary files.

File: .vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
+{
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "Debug Asterinas(debug)",
+            "type": "lldb", // install CodeLLDB extension first
```
Comment (tatetian) [inline .vscode/launch.json:6]: Interesting. So it seems that the VS code supports JSON format with comment extension.

File: .vscode/launch.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+            "request": "custom",
+            "targetCreateCommands": ["target create ${workspaceFolder}/target/x86_64-custom/debug/asterinas"],
+            "processCreateCommands": ["gdb-remote 1234"],
```
Comment (tatetian) [inline .vscode/launch.json:9]: Where does this magic number of 1234 come from?

Comment (fgh1999) [reply .vscode/launch.json:9]: `-s` equals to `-gdb tcp::1234`, i.e. open a gdbserver on TCP port 1234.

File: docs/src/kernel/advanced-instructions.md (hunk 2 above)
Comment (fgh1999) [reply docs/src/kernel/advanced-instructions.md]: The image is removed.

**Comment by junyang-zh** [comment]
Staying up late is bad, dude.

File: docs/src/kernel/advanced-instructions.md [hunk 5]
```diff
@@ -58,18 +62,32 @@ Then, in the interactive shell, run the following script to start the syscall te
...
 
 ```bash
-make run GDB_SERVER=1 ENABLE_KVM=0
+/asterinas/kernel# cargo osdk debug server
 ```
 
-Next, start a GDB client in another terminal.
+Then debug via [a GDB client in terminal](#terminal), or [VS Code](#vs-code).
+
+#### Terminal
```
Comment (junyang-zh) [inline docs/src/kernel/advanced-instructions.md:77]: Um, a little bit weird using headings here for me.

File: osdk/src/commands/debug.rs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            ]);
+            gdb.status().unwrap();
+        }
+        DebugRole::Server => {
+            vsc::config(workspace_root(), profile);
+
+            env::set_var(
+                "RUSTFLAGS",
+                env::var("RUSTFLAGS").unwrap_or_default()
+                    + " -g -C relocation-model=static -Z relro-level=off",
```
Comment (junyang-zh) [inline osdk/src/commands/debug.rs:42]: The "-C relocation-model=static -Z relro-level=off" flags are default for Asterinas now (Since, there is no KASLR, LOL). Adding these made them duplicated. And "-g" seems default for debug builds.

If there is indeed KASLR in the future we neither shouldn't change the relocation model here but to disable KASLR via kernel command line or a feature gate.

See https://github.com/asterinas/asterinas/blob/4eff679eacb2415b3e3077d6c0d261e742d87f23/osdk/src/commands/build/mod.rs#L127-L136. And [our discussions](https://github.com/asterinas/asterinas/pull/679#discussion_r1524660729)

File: osdk/src/config_manager/mod.rs [hunk 7]
```diff
@@ -69,6 +69,41 @@ impl RunConfig {
...
+    pub cargo_args: CargoArgs,
+    pub role: DebugRole,
+}
+
+impl DebugConfig {
+    pub fn parse(args: &DebugArgs) -> Self {
+        let cargo_args = split_features(&args.cargo_args);
+        let mut manifest = load_osdk_manifest(&cargo_args, args.osdk_args.select.as_ref());
+        apply_cli_args(&mut manifest, &args.osdk_args);
+        try_fill_system_configs(&mut manifest);
```
Comment (junyang-zh) [inline osdk/src/config_manager/mod.rs:90]: These four lines seems consistent across many Config parsing functions. Maybe we should make it them traits.

This is just an suggestion. I am neutral with the change though.

**Review by junyang-zh — CHANGES_REQUESTED** [review]
Nice improvement! Finally getting our GDB scripts back alive. Adding back those previously existed nasty GDB scripts for the firmware, bootloader or setup code should be taken by me.

The overall user's interface is OK with me, but here's some minor problems below.

File: docs/src/kernel/advanced-instructions.md (hunk 5 above)
Comment (fgh1999) [reply docs/src/kernel/advanced-instructions.md]: Now in list instead.

File: osdk/src/commands/debug.rs (hunk 6 above)
Comment (fgh1999) [reply osdk/src/commands/debug.rs]: The duplicate RUSTFLAGS options were added in case of forgetting to disable relocation after the link failures are resolved. They are removed now.

Sometimes I debug on a release build. It is much faster than the one in debug profile, especially on a machine without kvm.

File: osdk/src/config_manager/mod.rs (hunk 7 above)
Comment (fgh1999) [reply osdk/src/config_manager/mod.rs]: I'm not that clear about the expectation...
I just wrap the last three lines into a `get_final_manifest` method, which results in:

```rust
impl BuildConfig {
    pub fn parse(args: &BuildArgs) -> Self {
        let cargo_args = split_features(&args.cargo_args);
        Self {
            manifest: get_final_manifest(&cargo_args, &args.osdk_args),
            cargo_args,
        }
    }
}
```

**Review by junyang-zh — APPROVED** [review]
Looks good to me now.

**Comment by junyang-zh** [comment]
I had an informal conversation with Tate and we suggested some choices stabilizing the user's debugging interface of OSDK:
 - instead of `cargo osdk debug server`, `cargo osdk run --gdb-server` provide a better view since it emphasizes that we are running the kernel; this requires moving the server role code to `execute_run_command` and add some options for the run config;
 - the vsc feature should be optional rather than default; the user should do `cargo osdk run --gdb-server --vsc` to generate the json profile; and the created profile should be deleted when the server exits; add the profile to `.gitignore as well`.
 - the launching of clients should remain in the debug subcommand, like `cargo osdk debug`; expose the optional port/socket options: `cargo osdk debug --server localhost:1234`/`cargo osdk debug --server unix:.debug`; the unix socket should be used as default (since someone develop sharing the same machine).

Feel free to raise questions of the proposed user interface!

**Comment by fgh1999** [comment]
> (From doc) After starting a debug server with OSDK from the shell,
a `launch.json` will be generated/updated under `.vscode` if the directory exists.

The vsc feature for gdbserver utils does not generate/update a `launch.json` if there is no `.vscode` under the workspace root.
Thus, I want to confirm whether `--vsc` is necessary?
If the option is necessary, a detection of  `.vscode` will be removed.

>  and the created profile should be deleted when the server exits;

Sry, I was not so eager to handle such situation when the user had already gotten a launch config.
So,  if there exists a `.vscode/launch.json`, 
- resume its content to the state before the gdbserver started;
- or, just leave the new debug option in the original `launch.json`.

Which one is more acceptable?

> cargo osdk debug --server localhost:1234/cargo osdk debug --server unix:.debug

For the argument, I'd prefer`--remote` than `--server` for better consistence with other config manners, like `target remote` in gdb, or the 'remote' field in `launch.json`. Is it okay?

**Comment by junyang-zh** [comment]
Good points!
`--remote` is better than `--server`, I aggree.

The reason I want to make the vsc generation optional(at least, default but can be switched off) is that the developer should be free from worrying about unsupervised workspace file changes, since these files are tracked by git, and someone may accidentally stash these changes.

Since we have no tracked `launch.json`  in the repo now (and it seems not valid for a generated and tracked launch.json). Also, the launch.json seems not valid when the gdbserver is not running. I suggest just delete it if we created or modified it.

File: Makefile [hunk 8]
```diff
@@ -130,6 +130,14 @@ tools:
 run: build
 	@cd kernel && cargo osdk run $(CARGO_OSDK_ARGS)
 
+.PHONY: gdbserver
+gdbserver: build
```
Comment (tatetian) [inline Makefile:134]: `gdb_server` and `gdb_client` would be a bit easier on the eyes.

File: docs/src/kernel/advanced-instructions.md [hunk 9]
```diff
@@ -2,26 +2,30 @@
...
 
 ### Kernel-Mode Unit Tests
 
-Many crates in Asterinas do require a VM environment to be tested. The unit tests for these crates are empowered by our [ktest](framework/libs/ktest) framework.
+Many crates in Asterinas do require a VM environment to be tested.
+The unit tests for these crates are empowered by OSDK.
 
 ```bash
-make run KTEST=1
+/asterinas$ make ktest
```
Comment (tatetian) [inline docs/src/kernel/advanced-instructions.md:22]: I know what you mean by `<path>$ <command>`. But this syntax is non-standard. And for most of the time, the path is simply the root of the project. Thus, it is more of a distraction than an assistance. So I suggest removing `<path>$` from the commands and write in plain English to emphasize when the current working directory should be changed.

File: docs/src/kernel/advanced-instructions.md [hunk 10]
```diff
@@ -2,26 +2,30 @@
...
 ```
 
 It is also possible to specify a subset of tests to run.
 
 ```bash
-make run KTEST=1 KTEST_WHITELIST=failing_assertion,aster_frame::test::expect_panic KTEST_CRATES=aster-frame
+/asterinas/framework/aster-frame$ cargo osdk test failing_assertion
```
Comment (tatetian) [inline docs/src/kernel/advanced-instructions.md:28]: It is not entire obvious that `failing_assertion` is the name of a test case. And the description above says that `OSDK` support running a subset of tests. So the example command would be better of running multiple tests.

File: osdk/README.md [hunk 11]
```diff
@@ -76,6 +76,7 @@ Currently we support following commands:
 - **new**: Create a new kernel package or library package
 - **build**: Compile the project and its dependencies
 - **run**: Run the kernel with a VMM
+- **debug**: Debug a remote target via GDB
```
Comment (tatetian) [inline osdk/README.md:79]: Could you also add documentation about the debug support of OSDK in the Asterinas Book? Specifically, `OSDK User Reference > Commands > cargo osdk run` and `OSDK User Reference > Commands > cargo osdk debug`.

File: osdk/src/cli.rs [hunk 12]
```diff
@@ -109,6 +115,48 @@ pub struct RunArgs {
...
+#[derive(Debug, Args, Clone, Default)]
+pub struct GdbServerArgs {
+    /// Whether to enable QEMU GDB server for debugging
+    #[arg(
+        long,
+        short = 'G',
+        help = "Enable QEMU GDB server for debugging",
+        default_value_t
+    )]
+    pub gdb_server: bool,
```
Comment (tatetian) [inline osdk/src/cli.rs:131]: This name does not look at a boolean value. How about `is_gdb_enabled`?

File: osdk/src/cli.rs [hunk 13]
```diff
@@ -109,6 +115,48 @@ pub struct RunArgs {
...
     pub osdk_args: OsdkArgs,
+    #[command(flatten)]
+    pub gdb_server_args: GdbServerArgs,
+}
+
+#[derive(Debug, Args, Clone, Default)]
+pub struct GdbServerArgs {
+    /// Whether to enable QEMU GDB server for debugging
+    #[arg(
+        long,
```
Comment (tatetian) [inline osdk/src/cli.rs:126]: `--gdb-server`

File: osdk/src/cli.rs [hunk 14]
```diff
@@ -109,6 +115,48 @@ pub struct RunArgs {
...
+    pub gdb_server: bool,
+    #[arg(
+        long = "vsc",
+        help = "Generate a '.vscode/launch.json' for debugging with Visual Studio Code \
+                (only works when '--gdb-server' is enabled)",
+        default_value_t
+    )]
+    pub vsc_launch_file: bool,
+    #[arg(
+        long = "gdb-stub",
```
Comment (tatetian) [inline osdk/src/cli.rs:140]: I don't believe stub is the right word. What about `gdb-server-addr`?

File: osdk/src/cli.rs [hunk 15]
```diff
@@ -109,6 +115,48 @@ pub struct RunArgs {
...
+        help = "Path of GDB server stub",
```
Comment (tatetian) [inline osdk/src/cli.rs:141]: "The network address on which the GDB server listens. The address can be either a path for the UNIX domain socket or an IP address."

File: osdk/src/cli.rs [hunk 16]
```diff
@@ -109,6 +115,48 @@ pub struct RunArgs {
...
+        value_name = "PATH",
+        default_value = "/tmp/aster-gdb-socket"
```
Comment (tatetian) [inline osdk/src/cli.rs:143]: The default path should be a relative path to the Rust project root. Consider a OSDK-managed Rust workspace.

```
my-fancy-os-rs/
    Cargo.toml
    OSDK.toml
    my-fancy-kernel/
        Cargo.toml
        OSDK.toml
        src/
             lib.rs
    my-fancy-lib/
        Cargo.toml
        OSDK.toml
        src/
             lib.rs
```

The `cargo osdk run --gdb-server` command could be run under `my-fancy-os-rs/`, `my-fancy-kernel/`, or `my-fancy-lib/`. So in theory, there can be three OSDK-managed, GDB-enabled VMs running three different kernels at the same time. No to mention that there could be other OSDK-managed Rust projects being developed and tested on the same machine. So every file that `OSDK` generates or uses should be per-crate.

File: osdk/src/commands/launch.json.template [hunk 17]
```diff
@@ -0,0 +1,12 @@
+{
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "Debug Asterinas(#PROFILE#)",
+            "type": "lldb",
+            "request": "custom",
+            "targetCreateCommands": ["target create ${workspaceFolder}/target/x86_64-unknown-none/#PROFILE#/#BIN_NAME#"],
+            "processCreateCommands": ["gdb-remote 1234"]
```
Comment (tatetian) [inline osdk/src/commands/launch.json.template:9]: Shouldn't `1234` be replaced with the real server address?

File: osdk/src/commands/run.rs [hunk 18]
```diff
@@ -9,19 +9,80 @@ use super::{build::create_base_and_build, util::DEFAULT_TARGET_RELPATH};
...
     let default_bundle_directory = osdk_target_directory.join(target_name);
     let existing_bundle = Bundle::load(&default_bundle_directory);
 
+    let config = RunConfig {
+        manifest: {
+            if config.gdb_server_args.gdb_server {
+                let qemu_gdb_args: Vec<_> = {
+                    let gdb_stub_path = config.gdb_server_args.gdb_stub.as_str();
+                    match gdb::get_stub_type(gdb_stub_path) {
+                        gdb::StubType::Socket => {
```
Comment (tatetian) [inline osdk/src/commands/run.rs:38]: Technically, the Tcp option also uses sockets. So `StubType::Socket` should be `StubType::Unix`, which is short for UNIX sockets.

**Review by tatetian — COMMENTED** [review]
The changes look good for the most part. Could you also add some basic unit tests for the debug command?

**Comment by fgh1999** [comment]
As for unit tests, does anyone know a convient way to generate a ready-to-build kernel project? Can `cargo osdk new` one with some options?

**Comment by junyang-zh** [comment]
> As for unit tests, does anyone know a convient way to generate a ready-to-build kernel project? Can `cargo osdk new` one with some options?

You can refer to the integration tests written in the `osdk/tests` directory. The `cargo osdk new`ed project is ready to be built.

**Comment by fgh1999** [comment]
GDB needs to be added into the `osdk` image to pass osdk-doc-env-test, ref [8532a83](https://github.com/asterinas/asterinas/pull/626/commits/8532a83a6943f2475cfe936634bcf733a262df47)(to be modified).
I'm not sure about the difference between osdk-doc-env-test and osdk-test. Currently, they are doing quite similiar stuffs in different containers.

The changes related to docker images is in osdk/tools/Dockerfile.ubuntu22.04.
```bash
RUN apt update \ 
    && apt install -y \ 
    gdb \                # --- Here! ---
```

I want to confirm whether the `VERSION` should be shared between OSDK and Asterinas kernel? And, the docker build workflow appears not to update and push `asterinas:osdk`. Is it preferred to add a workflow for updating OSDK's docker image?

**Comment by junyang-zh** [comment]
> I want to confirm whether the `VERSION` should be shared between OSDK and Asterinas kernel?

Yes OSDK and the Asterinas project shares the version number.

> And, the docker build workflow appears not to update and push `asterinas/osdk`. Is it preferred to add a workflow for updating OSDK's docker image?

I think we should modify the `docker_push.yml` workflow to push the osdk docker image. Should be confirmed with @StevenJiang1110. But you are free to decide not to do it in this PR.

**Comment by fgh1999** [comment]
A "build & push" job is added in `osdk_publish.yml`.

File: .github/workflows/osdk_publish.yml [hunk 19]
```diff
@@ -21,3 +21,34 @@ jobs:
           args: --no-verify
           registry-token: ${{ secrets.CARGO_REGISTRY_TOKEN }}
           ignore-unpublished-changes: true
+
+  osdk-publish-dockerhub:
```
Comment (junyang-zh) [inline .github/workflows/osdk_publish.yml:25]: I suggest doing it in the `docker_build.yml` since:
 - you can reuse the fetch versions code there;
 - seems more consistent since the docker image is only used for testing and this image is not designed to be used by the OSDK user.

File: Makefile [hunk 20]
```diff
@@ -132,6 +132,14 @@ tools:
 run: build
 	@cargo osdk run $(CARGO_OSDK_ARGS)
 
+.PHONY: gdb_server
+gdb_server: build
+	@cd kernel && cargo osdk run $(CARGO_OSDK_ARGS) -G --vsc --gdb-stub :1234
```
Comment (junyang-zh) [inline Makefile:137]: Seems inconsistent: in the code and the doc the options is `--gdb-server-addr` but `--gdb-stub` here.

**Review by junyang-zh — COMMENTED** [review]
👀

File: Makefile (hunk 20 above)
Comment (fgh1999) [reply Makefile]: Sry. Forgot to adapt the cli interface...

File: .github/workflows/osdk_publish.yml (hunk 19 above)
Comment (fgh1999) [reply .github/workflows/osdk_publish.yml]: Moved to `docker_build.yml`.

File: .github/workflows/docker_build.yml [hunk 21]
```diff
@@ -29,7 +29,7 @@ jobs:
           echo "aster_version=$( cat VERSION )" >> "$GITHUB_OUTPUT"
           echo "rust_version=$( grep -m1 -o 'nightly-[0-9]\+-[0-9]\+-[0-9]\+' rust-toolchain.toml )" >> "$GITHUB_OUTPUT"
 
-      - name: Build and push
+      - name: Build and push kernel
```
Comment (junyang-zh) [inline .github/workflows/docker_build.yml:32]:
```suggestion
      - name: Build and push development image
```

File: .github/workflows/docker_build.yml [hunk 22]
```diff
@@ -39,3 +39,14 @@ jobs:
           tags: asterinas/asterinas:${{ steps.fetch-versions.outputs.aster_version }}
           build-args: |
             "ASTER_RUST_VERSION=${{ steps.fetch-versions.outputs.rust_version }}"
+
+      - name: Build and push OSDK
```
Comment (junyang-zh) [inline .github/workflows/docker_build.yml:43]:
```suggestion
      - name: Build and push OSDK test image
```

**Review by junyang-zh — COMMENTED** [review]
Two minor fixes and we are good to go!

Maybe merge this PR first before #704. Rebasing the former one to this is easier.

**Comment by tatetian** [comment]
@fgh1999 Could you take a look at the CI failure?

**Comment by junyang-zh** [comment]
Maybe we could merge it now.

I am in need of this feature. QaQ

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!
